### PR TITLE
Bg 1848 20 char limit

### PR DESCRIPTION
**Pull request by brick-green-agile6**
This PR filters the usage of HD_TRUNCATION_LIMIT to account for HL7 special characters presetn in the incoming value. Special HL7 characters, when sent as part of a string input value, have to be escaped. That escaping adds additional characters to the string causing the 20 character limit to be exceeded. This update will first check if special characters are present then adjust the truncation limit accordingly..

## Changes
- added function setTruncationLimitWithEncoding() to HL7Serializer to find special characters in value then shorten the truncation limit
- set function in HL7 serializer

### Testing
- [X] Tested locally?
- [X] Ran `quickTest all`?
- [X] Ran `./prime test` against local Docker ReportStream container?
- [X] Downloaded a file from `http://localhost:7071/api/download`?
- [ ] Added tests?

## Fixes
*List GitHub issues this PR fixes*
- #1848

### CODE REVIEW

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt [hunk 1]
```diff
@@ -924,6 +924,16 @@ class Hl7Serializer(val metadata: Metadata) : Logging {
         terser.set("/PATIENT_RESULT/ORDER_OBSERVATION/OBSERVATION/OBX-23-7", "XX")
     }
 
+    private fun setTruncationLimitWithEncoding(value: String, truncationLimit: Int): Int {
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:927]: Need kdoc

Comment (carlosfelix2) [reply prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:927]: Add unit tests for this method.  What happens with 0, 1 and many escape characters (not sure if many are allowed).

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt [hunk 2]
```diff
@@ -924,6 +924,16 @@ class Hl7Serializer(val metadata: Metadata) : Logging {
...
+        val regex = "[^&~|]".toRegex()
+        val match = regex.containsMatchIn(value)
+        return if (match) {
+            truncationLimit.minus(2)
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:931]: Can there be multiple escape characters?

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt (hunk 2 above)
Comment (brick-green-agile6) [reply prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:931]: Yep, there can be multiple and this wouldn't account for them all. Good catch! I'll adjust

File: prime-router/src/test/kotlin/serializers/Hl7SerializerTests.kt [hunk 3]
```diff
@@ -635,6 +636,16 @@ NTE|1|L|This is a final comment|RE"""
         }
     }
 
+    @Test
+    fun `test setTruncationLimitWithEncoding`() {
+
+        val testValue = "Test & Value ~ Text ^ String"
```
Comment (carlosfelix2) [inline prime-router/src/test/kotlin/serializers/Hl7SerializerTests.kt:642]: When doing unit tests, you need to cover all scenarios.  In this case you are missing the no special characters scenario to cover the else in the if statement.

Comment (brick-green-agile6) [reply prime-router/src/test/kotlin/serializers/Hl7SerializerTests.kt]: got this fixed

**Review by carlosfelix2 — APPROVED** [review]

File: prime-router/src/test/kotlin/serializers/Hl7SerializerTests.kt [hunk 4]
```diff
@@ -635,6 +636,19 @@ NTE|1|L|This is a final comment|RE"""
         }
     }
 
+    @Test
+    fun `test setTruncationLimitWithEncoding`() {
+
+        val testValueWithSpecialChars = "Test & Value ~ Text ^ String"
+        val testValueNoSpecialChars = "Test Value Text String"
+        val testLimit = 20
```
Comment (carlosfelix2) [inline prime-router/src/test/kotlin/serializers/Hl7SerializerTests.kt:644]: Don't feel you have to define all the numbers separately.  next time you could do just
        assertEquals(serializer.getTruncationLimitWithEncoding(testValueWithSpecialChars, testLimit), 14)

**Pull request merged**